subprojects/model-core/src/main/java/org/gradle/api/internal/model/NamedObjectInstantiator.java

```diff
@@ -54,17 +59,38 @@
     private static final String[] EMPTY_STRINGS = new String[0];
     private static final String CONSTRUCTOR_NAME = "<init>";
 
-    // Currently retains strong references to types
-    private final LoadingCache<Class<?>, LoadingCache<String, Object>> values = CacheBuilder.newBuilder().build(new CacheLoader<Class<?>, LoadingCache<String, Object>>() {
+    private final Factory<LoadingCache<Class<?>, LoadingCache<String, Object>>> cacheFactory = new Factory<LoadingCache<Class<?>, LoadingCache<String, Object>>>() {
         @Override
-        public LoadingCache<String, Object> load(Class<?> type) {
-            return CacheBuilder.newBuilder().build(loaderFor(type));
+        public LoadingCache<Class<?>, LoadingCache<String, Object>> create() {
+            return newValuesCache();
         }
-    });
+    };
+
+    // Currently retains strong references to types
+    private final LoadingCache<Class<?>, LoadingCache<String, Object>> leakyValues = newValuesCache();
+
+    private LoadingCache<Class<?>, LoadingCache<String, Object>> newValuesCache() {
+        return CacheBuilder.newBuilder()
+                .weakKeys()
```

Comment:
❌ Weak keys aren't necessary here (and won't help either)

Reply:
Indeed, leftover of previous implementation.